c/shared/source/uforead/uforead.c

```diff
@@ -1819,9 +1818,153 @@ static void reallocFDArray(ufoCtx h){
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        keyID = NULL;
+        if ((xmlStrEqual(cur->name, (const xmlChar *) "key"))) {
+            parseKeyName(&keyID, cur);
+            if (keyID == NULL){
+                message(h, "Warning: Encountered empty <key></key>.");
+                cur = cur->next;
+                continue;
+            }
+            keyName = (char*) keyID;
+            cur = cur->next;
+            setFontDictKey(h, keyName, cur);
+        }
+        cur = cur->next;
+    }
+}
+
+static void parseKeyName(unsigned char* *keyID, xmlNodePtr cur){
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        if (!xmlStrEqual(cur->name, (const xmlChar *) "text"))
+            continue;
+        *keyID = xmlNodeGetContent(cur);
+        cur = cur->next;
+    }
+}
+
+static void parseXMLArray(ufoCtx h, xmlNodePtr cur){
+    dnaSET_CNT(h->valueArray, 0);
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        void* valueString = parseKeyValue(h, cur);
+        if (valueString != NULL)
+            *dnaNEXT(h->valueArray) = valueString;
+        cur = cur->next;
+    }
+}
+
+static void parseXMLDict(ufoCtx h, xmlNodePtr cur){
+    // go through an XML dict
+    unsigned char* keyID;
+    cur = cur->xmlChildrenNode;
+
+    if (parsingFDArray){
+        currentiFD = currentiFD + 1;
+        h->top.FDArray.cnt = h->top.FDArray.cnt + 1;
+        if (h->top.FDArray.cnt > FDArrayInitSize){ /* Memory needs reallocation*/
+            reallocFDArray(h);
+        }
+        abfFontDict* fd = h->top.FDArray.array + currentiFD;
+        abfInitFontDict(fd);
+    }
+
+    while (cur != NULL) {
+        parseKeyName(&keyID, cur);
+//        cur = cur->next;
+//        void *ptr = parseKeyValue(h, cur);
+
+        char* keyName = (char*) keyID;
+        cur = cur->next;
+        setFontDictKey(h, keyName, cur);
+        cur = cur->next;
+    }
+}
+
+static bool isSimpleKey(xmlNodePtr cur){
+    if (xmlStrEqual(cur->name, (const xmlChar *) "string" ) ||
+        xmlStrEqual(cur->name, (const xmlChar *) "integer") ||
+        xmlStrEqual(cur->name, (const xmlChar *) "real")    ||
+        xmlStrEqual(cur->name, (const xmlChar *) "date")) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+static char* parseKeyError(ufoCtx h, xmlNodePtr cur){
+    if (xmlStrEqual(cur->name, (const xmlChar *) "key"))
+        fatal(h, ufoErrParse, "Encountered empty <key>");  // needs fix
+    else if (xmlStrEqual(cur->name, (const xmlChar *) "array"))
+        fatal(h, ufoErrParse, "Encountered empty <array>");
+    else
+        return NULL;
+}
+
+static void *parseKeyValue(ufoCtx h, xmlNodePtr cur){
+    if (isSimpleKey(cur)) {  /* if string, integer, or real */
+        void *ptr = xmlNodeGetContent(cur);
+        return ptr;
+    } else if (xmlStrEqual(cur->name, (const xmlChar *) "dict")) {
+        parseXMLDict(h, cur);
```

Comment:
We never define what we're returning here, as far as I can see. We just slip through the rest of the guards to the end of the function.